src/Data/Functor/Classes/Generic.hs

```diff
@@ -0,0 +1,154 @@
+genericLiftShowsPrec sp sl d = gliftShowsPrec sp sl d . from1
+
+
+-- Generics
+
+instance GEq1 U1 where
+  gliftEq _ _ _ = True
+
+instance GEq1 Par1 where
+  gliftEq f (Par1 a) (Par1 b) = f a b
+
+instance Eq c => GEq1 (K1 i c) where
+  gliftEq _ (K1 a) (K1 b) = a == b
+
+instance Eq1 f => GEq1 (Rec1 f) where
+  gliftEq f (Rec1 a) (Rec1 b) = liftEq f a b
+
+instance GEq1 f => GEq1 (M1 i c f) where
+  gliftEq f (M1 a) (M1 b) = gliftEq f a b
+
+instance (GEq1 f, GEq1 g) => GEq1 (f :+: g) where
+  gliftEq f a b = case (a, b) of
+    (L1 a, L1 b) -> gliftEq f a b
+    (R1 a, R1 b) -> gliftEq f a b
+    _ -> False
+
+instance (GEq1 f, GEq1 g) => GEq1 (f :*: g) where
+  gliftEq f (a1 :*: b1) (a2 :*: b2) = gliftEq f a1 a2 && gliftEq f b1 b2
+
+instance (Eq1 f, GEq1 g) => GEq1 (f :.: g) where
+  gliftEq f (Comp1 a) (Comp1 b) = liftEq (gliftEq f) a b
+
+
+instance GOrd1 U1 where
+  gliftCompare _ _ _ = EQ
+
+instance GOrd1 Par1 where
+  gliftCompare f (Par1 a) (Par1 b) = f a b
+
+instance Ord c => GOrd1 (K1 i c) where
+  gliftCompare _ (K1 a) (K1 b) = compare a b
+
+instance Ord1 f => GOrd1 (Rec1 f) where
+  gliftCompare f (Rec1 a) (Rec1 b) = liftCompare f a b
+
+instance GOrd1 f => GOrd1 (M1 i c f) where
+  gliftCompare f (M1 a) (M1 b) = gliftCompare f a b
+
+instance (GOrd1 f, GOrd1 g) => GOrd1 (f :+: g) where
+  gliftCompare f a b = case (a, b) of
+    (L1 a, L1 b) -> gliftCompare f a b
+    (R1 a, R1 b) -> gliftCompare f a b
+    (L1 _, R1 _) -> LT
+    (R1 _, L1 _) -> GT
+
+instance (GOrd1 f, GOrd1 g) => GOrd1 (f :*: g) where
+  gliftCompare f (a1 :*: b1) (a2 :*: b2) = gliftCompare f a1 a2 <> gliftCompare f b1 b2
+
+instance (Ord1 f, GOrd1 g) => GOrd1 (f :.: g) where
+  gliftCompare f (Comp1 a) (Comp1 b) = liftCompare (gliftCompare f) a b
+
+
+instance GShow1 Par1 where
+  gliftShowsPrec sp _ d (Par1 a) = sp d a
+
+instance Show c => GShow1 (K1 i c) where
+  gliftShowsPrec _ _ d (K1 a) = showsPrec d a
+
+instance Show1 f => GShow1 (Rec1 f) where
+  gliftShowsPrec sp sl d (Rec1 a) = liftShowsPrec sp sl d a
+
+instance GShow1 f => GShow1 (M1 D c f) where
+  gliftShowsPrec sp sl d (M1 a) = gliftShowsPrec sp sl d a
+
+instance (Constructor c, GShow1Body f) => GShow1 (M1 C c f) where
+  gliftShowsPrec sp sl d m = gliftShowsPrecBody (conFixity m) (conName m) sp sl d (unM1 m)
+
+instance GShow1Body U1 where
+  gliftShowsPrecBody _ conName _ _ _ _ = showString conName
+
+instance GShow1 f => GShow1Body (M1 S c f) where
+  gliftShowsPrecBody _ conName sp sl d (M1 a) = showsUnaryWith (gliftShowsPrec sp sl) conName d a
+
+instance (GShow1 f, GShow1 g) => GShow1Body (f :*: g) where
+  gliftShowsPrecBody fixity conName sp sl d (a :*: b) = case fixity of
+    Prefix       -> showsUnaryWith (gliftShowsPrec sp sl) conName d a
+    Infix _ prec -> showParen (d > prec) $ gliftShowsPrec sp sl (succ prec) a . showChar ' ' . showString conName . showChar ' ' . gliftShowsPrec sp sl (succ prec) b
```

Comment:
I’m pretty happy with how this bit turned out. Handles infix constructors just fine.